test/e2e/external_oidc_test.go

```diff
@@ -187,263 +290,460 @@ func newTestClient(t *testing.T) (*testClient, error) {
+		return nil, nil, fmt.Errorf("could not get issuer OpenID well-known JWKS configuration: %v", err)
 	}
+	defer resp.Body.Close()
 
-	proxy, err := tc.configClient.ConfigV1().Proxies().Get(ctx, "cluster", metav1.GetOptions{})
+	respBytes, err = io.ReadAll(resp.Body)
 	if err != nil {
-		return
+		return nil, nil, fmt.Errorf("could not parse well-known JWKS response body: %v", err)
 	}
 
-	origTrustedCAName := proxy.Spec.TrustedCA.Name
-	proxy.Spec.TrustedCA.Name = "default-ingress-cert"
-	proxy, err = tc.configClient.ConfigV1().Proxies().Update(ctx, proxy, metav1.UpdateOptions{})
-	cleanups = append(cleanups, func() {
-		proxy.Spec.TrustedCA.Name = origTrustedCAName
-		proxy, err = tc.configClient.ConfigV1().Proxies().Update(ctx, proxy, metav1.UpdateOptions{})
-		if err != nil {
-			tc.t.Fatalf("cleanup failed for proxy '%s': %v", proxy.Name, err)
+	type jwks struct {
+		Keys []struct {
+			KID string `json:"kid"`
+			Use string `json:"use"`
+			KTY string `json:"kty"`
+			Alg string `json:"alg"`
+			N   string `json:"n"`
+			E   string `json:"e"`
+		} `json:"keys"`
+	}
+
+	var issuerJWKS jwks
+	if err := json.Unmarshal(respBytes, &issuerJWKS); err != nil {
+		return nil, nil, fmt.Errorf("could not unmarshal JWKS: %v", err)
+	}
+
+	keyfunc := func(token *jwt.Token) (any, error) {
+		for _, key := range issuerJWKS.Keys {
+			if key.KID == token.Header["kid"] {
+				switch key.Alg {
+				case "RS256":
+					n, err := base64.RawURLEncoding.DecodeString(key.N)
+					if err != nil {
+						return nil, fmt.Errorf("could not decode N: %v", err)
+					}
+					e, err := base64.RawURLEncoding.DecodeString(key.E)
+					if err != nil {
+						return nil, fmt.Errorf("could not decode E: %v", err)
+					}
+
+					pubkey := &rsa.PublicKey{
+						N: new(big.Int).SetBytes(n),
+						E: int(new(big.Int).SetBytes(e).Int64()),
+					}
+
+					return pubkey, nil
+				}
+
+				return nil, fmt.Errorf("unexpected signing algorithm for key '%s': %s", key.KID, key.Alg)
+			}
 		}
-	})
+
+		return nil, fmt.Errorf("could not find an RSA key for signing use in the provided JWKS")
+	}
+
+	parsedAccessToken, err := jwt.ParseWithClaims(accessToken, &expectedClaims{}, keyfunc)
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not parse access token: %v", err)
+	}
+
+	parsedIDToken, err := jwt.ParseWithClaims(idToken, &expectedClaims{}, keyfunc)
 	if err != nil {
-		return
+		return nil, nil, fmt.Errorf("could not parse ID token: %v", err)
 	}
 
-	return
+	return parsedAccessToken, parsedIDToken, nil
+}
+
+func (tc *testClient) setupKeycloakClient(t *testing.T, ctx context.Context, keycloakURL string) *test.KeycloakClient {
+	transport, err := rest.TransportFor(tc.kubeConfig)
+	require.NoError(t, err)
+
+	kcClient := test.KeycloakClientFor(t, transport, keycloakURL, "master")
+	err = wait.PollUntilContextTimeout(ctx, 5*time.Second, 30*time.Second, true, func(ctx context.Context) (bool, error) {
+		err := kcClient.AuthenticatePassword(oidcClientId, "", "admin", "password")
```

Comment:
We have a way to use an existing keycloak deployment using an environment variable when running this e2e test - what happens if `admin` and `password` are not the right login credentials for that existing deployment?

Should this also be configurable in the same way? 

Alternatively, should we just always stand up a new Keycloak instance for this test?

Reply:
Using an existing instance for the test is only meant to aid development (as we'll be able to run the test immediately instead of creating and waiting for the new keycloak to come up). The primary mode of operation of the test will be to create a new instance using `keycloakidp.go`.

Reply:
While I see the value from the devex perspective, is the keycloak instance cleaned up after the test ends in the primary mode of operation?

I am a bit hesitant to have this injection mode that could potentially be abused beyond devex and cause confusion. IMO we either go fully configurable to use an existing keycloak instance or not at all.

Reply:
How much time is actually saved per run by not having to stand up keycloak?

Reply:
In the primary mode of operation, the keycloak instance is cleaned up; its whole namespace gets deleted. There's another env var (`OPENSHIFT_KEEP_IDP`) that is used to prevent the cleanup, for debugging purposes.

It's not so much about how much time we save, but rather sometimes there can be failures associated with the keycloak being deployed successfully which makes it frustrating when you're trying to test something else. This was mainly useful during the bulk of the development process, so I'm happy to drop this now. It's easy to quickly modify the test in this way again anyway.